library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/brewing/CustomPotionBrewingRecipe.java

```
@@ -0,0 +1,191 @@
+
+		var statusEffects = this.statusEffects;
+		if (input.getOrCreateNbt().contains("CustomPotionEffects")) {
+			statusEffects = new ArrayList<>(this.statusEffects);
+			statusEffects.addAll(PotionUtil.getCustomPotionEffects(input));
+		}
+
+		PotionUtil.setCustomPotionEffects(output, statusEffects);
+
+		return output;
+	}
+
+	@Override
+	public RecipeSerializer<CustomPotionBrewingRecipe> getSerializer() {
+		return RecipeImpl.CUSTOM_POTION_SERIALIZER;
+	}
+
+	public static class Serializer extends PotionBrewingRecipe.Serializer<CustomPotionBrewingRecipe> {
+		public Serializer(RecipeFactory<Potion, CustomPotionBrewingRecipe> recipeFactory) {
+			super(recipeFactory);
+		}
+
+		@Override
+		public CustomPotionBrewingRecipe read(Identifier id, JsonObject json) {
+			CustomPotionBrewingRecipe recipe = super.read(id, json);
+
+			if (json.has("effects")) {
+				recipe.statusEffects.clear();
+				for(JsonElement effectJson : JsonHelper.getArray(json, "effects")) {
+					StatusEffectInstance instance;
+					instance = readStatusEffectInstance(effectJson);
+					recipe.statusEffects.add(instance);
+				}
+			}
+
+			return recipe;
+		}
+
+		private StatusEffectInstance readStatusEffectInstance(JsonElement json) {
+			if (json instanceof JsonObject jsonObject) {
+				String string = JsonHelper.getString(jsonObject, "type");
+				StatusEffect statusEffect = Registry.STATUS_EFFECT
+						.getOrEmpty(new Identifier(string))
+						.orElseThrow(() -> new JsonSyntaxException("Unknown status effect '" + string + "'"));
+				int duration = JsonHelper.getInt(jsonObject, "duration", 20);
+				int amplifier = JsonHelper.getInt(jsonObject, "amplifier", 0);
+				boolean showParticles = JsonHelper.getBoolean(jsonObject, "particles", true);
+				boolean showIcon = JsonHelper.getBoolean(jsonObject, "icon", true);
+				return new StatusEffectInstance(statusEffect, duration, amplifier, false, showParticles, showIcon);
+			} else { // default to a basic 20 tick 0 amplifier status effect
```

I'm worried this could cause situations where invalid jsons are silently ignored. Is that intended, and if so is it normal for Minecraft?

---

The idea is to have two possible JSON objects. Either a simple string for a default potion effect or a bigger object letting you define exact values.